sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonFilters.scala

```diff
@@ -0,0 +1,133 @@
+    val groupedPredicates = Array.fill(schema.length)(Array.empty[JsonPredicate])
+    if (SQLConf.get.jsonFilterPushDown) {
+      val groupedByRefSet = filters
+        // Group filters that have the same set of references. For example:
+        //   IsNotNull("i") -> Set("i"), AlwaysTrue -> Set(),
+        //   And(EqualTo("i", 0), StringStartsWith("s", "abc")) -> Set("i", "s")
+        // By grouping filters we could avoid tracking their state of references in the
+        // current row separately.
+        .groupBy(_.references.toSet)
+        // Combine all filters from the same group by `And` because all filters should
+        // return `true` to do not skip a row. The result is compiled to a predicate.
+        .map { case (refSet, refsFilters) =>
+          (refSet, JsonPredicate(toPredicate(refsFilters), refSet.size, 0))
+        }
+      // Apply predicates w/o references like AlwaysTrue and AlwaysFalse to all fields.
+      // We cannot set such predicates to a particular position because skipRow() can
+      // be invoked for any index due to unpredictable order of JSON fields in JSON records.
+      val withLiterals = groupedByRefSet.map { case (refSet, pred) =>
+        if (refSet.isEmpty) {
+          (schema.fields.map(_.name).toSet, pred.copy(totalRefs = 1))
+        } else {
+          (refSet, pred)
+        }
+      }
+      // Build a map where key is only one field and value is seq of predicates refer to the field
+      // "i" -> Seq(AlwaysTrue, IsNotNull("i"), And(EqualTo("i", 0), StringStartsWith("s", "abc")))
+      // "s" -> Seq(AlwaysTrue, And(EqualTo("i", 0), StringStartsWith("s", "abc")))
+      val groupedByFields = withLiterals.toSeq
+        .flatMap { case (refSet, pred) => refSet.map((_, pred)) }
+        .groupBy(_._1)
+      // Build the final array by converting keys of `groupedByFields` to their
+      // indexes in the provided schema.
+      groupedByFields.foreach { case (fieldName, fieldPredicates) =>
+        val fieldIndex = schema.fieldIndex(fieldName)
+        groupedPredicates(fieldIndex) = fieldPredicates.map(_._2).toArray
+      }
+    }
+    groupedPredicates
+  }
+
+  /**
+   * Applies predicates (compiled filters) associated with the row field value
+   * at the position `index` only if other predicates dependencies are already
+   * set in the given row.
+   *
+   * @param row The row with fully or partially set values.
+   * @param index The index of already set value.
+   * @return true if at least one of applicable predicates (all dependent row values are set)
+   *         return `false`. It returns `false` if all predicates return `true`.
+   */
+  def skipRow(row: InternalRow, index: Int): Boolean = {
+    var skip = false
+    predicates(index).foreach { pred =>
+      pred.refCount -= 1
+      assert(pred.refCount >= 0,
+        s"Predicate reference counter cannot be negative but got ${pred.refCount}.")
+      if (!skip && pred.refCount == 0) {
```

Comment:
nit: we can return earlier of `skip` becomes true.

Reply:
Right, we can but the state of refCounts will be inconsistent - some refCount were decremented but some not. There are tests that check the state. Let me try to modify the tests.